### PR TITLE
mon: stuck peering since warning is misleading

### PR DESCRIPTION
**Pull request by shreyanshjain7174**
When osds restart or manually marked down it is common to see a HEALTH_WARN claiming that PGs have been stuck peering since awhile, even though they were active. The warning is to be issued if they really are stuck peering longer than 60s.

Fixes: https://tracker.ceph.com/issues/51688

Signed-off-by: shreyanshjain7174 <ssanchet@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by shreyanshjain7174** [comment]
@kamoltat have a look at it!

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
jenkins test api

**Comment by kamoltat** [comment]
jenkins test windows

**Comment by kamoltat** [comment]
jenkins test make check arm64

File: src/mon/PGMap.cc
Comment (kamoltat) [inline src/mon/PGMap.cc:3860]: From what I understand, you're trying to update the last_peered state of the PG. However, this is already done in src/osd/PeeringState.cc. Therefore I think this block of code is not relevant.

**Review by kamoltat — CHANGES_REQUESTED** [review]
Left a comment regarding a possible redundancy in terms of updating last_peered.
Other than that this LGTM!

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
jenkins test make check arm64

File: src/mon/PGMap.cc
Comment (shreyanshjain7174) [reply src/mon/PGMap.cc:3860]: Will discard it and commit again.

**Comment by kamoltat** [comment]
jenkins test make check

**Review by kamoltat — APPROVED** [review]
LGTM!

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by rzarzynski** [comment]
@kamoltat: is the lack of test a blocker?

**Comment by dvanders** [comment]
@shreyanshjain7174 Sorry, I don't see how this fixes the original issue. (This code just adds peering state to pg dump output, afaict).

@rzarzynski yes lack of test is block -- we should not merge this until it can be reproduced properly.

@shreyanshjain7174 Via the code at https://github.com/ceph/ceph/blob/main/src/mon/PGMap.cc#L2434 we'll still get warnings for PGs that are peering but have been idle a long time.

Did you try reproducing using thg the method in the tracker?
> I tracked this to `PGMap::get_health_checks` which will mark a PG as stuck peering if now - last_peered > mon_pg_stuck_threshold.
But the problem is that last_peered is only updated if there is IO on a PG -- an OSD doesn't send pgstats if it is idle.
To fix, we could update last_active/last_peered etc, and send a pg stats update more frequently even when idle?

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by shreyanshjain7174** [comment]
@dvanders I will again try it with unit tests, I tried to reproduce with the method mentioned in the tracker and checked only for a min or so, didn't tried to check if the pgs are in peering state staying idle. I tried to bring osd node down and started it again in a minute or so. Let me try other cases. 
Thanks for the input.

**Comment by rzarzynski** [comment]
@pdvian: Hi Prashant! How about turning this reproducer (from comment no. 10 in the tracker) into a workunit and appending it to the PR?

**Comment by ljflores** [comment]
Rados suite review: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcCcKXkHLe1789-wip-yuri3-testing-2023-06-19-1518

**Review by ljflores — DISMISSED** [review]
@shreyanshjain7174 @rzarzynski is this one ready for merging? I'm reading past comments and it looks like we may be waiting on additional tests? Let me know and I'll remove this request for changes.

**Comment by yuriw** [comment]
ref: https://trello.com/c/CcKXkHLe

**Comment by shreyanshjain7174** [comment]
> @shreyanshjain7174 @rzarzynski is this one ready for merging? I'm reading past comments and it looks like we may be waiting on additional tests? Let me know and I'll remove this request for changes.

I am creating a workunit for this PR and going to add it in this PR only until then it shouldn't be merged.

**Comment by yuriw** [comment]
@shreyanshjain7174 @ljflores pls merge when ready

**Pull request merged**
